tools/continuous_batching/limit_checker.py

```diff
@@ -0,0 +1,247 @@
+
+
+        while True:
+            gc.collect()
+            generation_config = GenerationConfig()  # expecting default greedy sampling
+            generation_config.num_return_sequences = 1
+            generation_config.max_new_tokens = generation_length + 1
+            generation_config.apply_chat_template = False
+            generation_config.ignore_eos = True
+            print(f"generation_length:{generation_length} ", sep='')
+
+            start = time.time()
+            max_cache_usage = run_and_write_metrics(model_cb_opt, prompt, generation_config, args.report)
+            end = time.time()
+            if (end - start) > args.timeout:
+                print("Maximum generation time reached")
+                break
+            elif max_cache_usage == 100:
+                print("Cache size exhausted")
+                break
+
+            generation_length *= 2
+    elif args.mode == "gen_throughput":
+        dataset = load_samsum_dataset(args.data)
+        prompt_throughput = 1
+        prompt_left_bound = prompt_throughput
+        prompt_right_bound = None
+        is_right_bound = False
+
+        while True:
+            gc.collect()
+            generation_config = GenerationConfig()  # expecting default greedy sampling
+            generation_config.num_return_sequences = 1
+            generation_config.apply_chat_template = False
+            prompt_subset = dataset["prompts"][:prompt_throughput]
+            print(f"prompt_throughput {prompt_throughput}")
+            result: GenerationResult = model_cb_opt.generate(prompt_subset, generation_config=[generation_config] * len(prompt_subset))
+
+            pipeline_opt_metrics = model_cb_opt.get_metrics()
+            rss_usage_gb = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 3
+            print(f"avg_cache_usage:{pipeline_opt_metrics.avg_cache_usage:.2f}% max_cache_usage:{pipeline_opt_metrics.max_cache_usage:.2f}% rss_usage:{rss_usage_gb:.3f} GB")
+            print()
+
+            max_cache_usage = pipeline_opt_metrics.max_cache_usage
+
+            if max_cache_usage == 100.0 and not is_right_bound:
+                is_right_bound = True
+                prompt_right_bound = prompt_throughput
+
+            if not is_right_bound:
+                prompt_left_bound = prompt_throughput
+                prompt_throughput *= 2
+            else:
+                if max_cache_usage == 100.0:
+                    prompt_right_bound = prompt_throughput
+                elif max_cache_usage < 100.0:
+                    prompt_left_bound = prompt_throughput
+                prompt_throughput = (prompt_left_bound + prompt_right_bound) // 2
+
+                if (prompt_right_bound - prompt_left_bound <= 1):
+                    break
+
+
+        print(f"Approximate highest throughput: {prompt_throughput} prompts")
+        del data_dict
```

Comment:
Should this line be moved into the `if args.mode == "gen_length"` block?

Reply:
Done

Reply:
`del data_dict` is still at the end of the file